src/sinks/gcp/chronicle_unstructured.rs

```diff
@@ -0,0 +1,528 @@
+use bytes::{Bytes, BytesMut};
+use futures_util::{future::BoxFuture, task::Poll};
+use goauth::scopes::Scope;
+use http::{header::HeaderValue, Request, Uri};
+use hyper::Body;
+use indoc::indoc;
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use snafu::Snafu;
+use std::io;
+use tokio_util::codec::Encoder as _;
+use tower::{Service, ServiceBuilder};
+use vector_buffers::Ackable;
+use vector_core::{
+    config::{AcknowledgementsConfig, Input},
+    event::{Event, EventFinalizers, Finalizable},
+    sink::VectorSink,
+};
+
+use crate::{
+    config::{log_schema, GenerateConfig, SinkConfig, SinkContext, SinkDescription},
+    gcp::{GcpAuthConfig, GcpAuthenticator},
+    http::{HttpClient, HttpError},
+    sinks::{
+        gcs_common::{
+            config::{healthcheck_response, GcsRetryLogic},
+            service::GcsResponse,
+            sink::GcsSink,
+        },
+        util::{
+            encoding::{
+                Encoder, EncodingConfig, EncodingConfigWithFramingAdapter, StandardEncodings,
+                StandardEncodingsWithFramingMigrator,
+            },
+            metadata::{RequestMetadata, RequestMetadataBuilder},
+            partitioner::KeyPartitioner,
+            request_builder::EncodeResult,
+            BatchConfig,
+            BulkSizeBasedDefaultBatchSettings,
+            Compression,
+            //Compressor,
+            RequestBuilder,
+            TowerRequestConfig,
+        },
+        Healthcheck,
+    },
+    template::{Template, TemplateParseError},
+    tls::{TlsConfig, TlsSettings},
+};
+
+const NAME: &str = "gcp_chronicle_unstructured";
+
+
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub))]
+pub enum GcsHealthcheckError {
+    #[snafu(display("log_type template parse error: {}", source))]
+    LogTypeTemplate { source: TemplateParseError },
+
+    #[snafu(display("Endpoint not found"))]
+    NotFound,
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum Region {
+    Eu,
+    Us,
+    Asia,
+}
+
+impl Region {
+    /// Each region has a it's own endpoint.
+    fn endpoint(self) -> &'static str {
+        match self {
+            Region::Eu => "https://europe-malachiteingestion-pa.googleapis.com",
+            Region::Us => "https://malachiteingestion-pa.googleapis.com",
+            Region::Asia => "https://asia-southeast1-malachiteingestion-pa.googleapis.com",
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct ChronicleUnstructuredConfig {
+    pub endpoint: Option<String>,
+    pub region: Option<Region>,
+    pub customer_id: String,
+    #[serde(flatten)]
+    pub auth: GcpAuthConfig,
+    #[serde(default)]
+    pub batch: BatchConfig<BulkSizeBasedDefaultBatchSettings>,
+    #[serde(flatten)]
+    pub encoding: EncodingConfigWithFramingAdapter<
+        EncodingConfig<StandardEncodings>,
+        StandardEncodingsWithFramingMigrator,
+    >,
```

Comment:
@pablosichert Can I do this encoding better?

Reply:
You can (and should) use `EncodingConfigWithFraming` here instead 😄

Also see #13518, the `Adapter`s will be gone shortly 😄